### PR TITLE
Feature/475/export 3d printer files

### PR DESCRIPTION
**Pull request by IhsenBouallegue**
# Introduction of 3D printing feature

Close: #475 

## Description

Now we can export .stl files from visualization. These files are then used with a 3D slicer software to generate the appropriate G-code for the 3D printer.

With the current implementation, the opened map will be downloaded as an .stl file after clicking the export 3D button.

## How-to

1. Click the 3D export button
grafik

2. Save the STL file
grafik

3. Import file into your chosen slicer software and configure (Cura 3d shown below)
grafik

4. Print! 🖨️ 🥳 
grafik

### CODE REVIEW

**Review by Christian-Eberhard — APPROVED** [review]

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeSceneService.ts [hunk 1]
```diff
@@ -35,6 +35,8 @@ export class ThreeSceneService implements CodeMapPreRenderServiceSubscriber, Map
 	mapGeometry: Group
 	private readonly lights: Group
 	private mapMesh: CodeMapMesh
+	/** Temporary workaround until mapMesh can be derived from store */
```
Comment (ce-bo) [inline visualization/app/codeCharta/ui/codeMap/threeViewer/threeSceneService.ts:38]: The TODO Prefix will highlight this comment in IDEs :-)
```suggestion
	/** TODO: Fix temporary workaround as soon as mapMesh can be derived from store */
```

File: visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import { Mesh } from "three"
+
+@Component({
+	selector: "cc-export-threed-map-button",
+	template: require("./export3DMapButton.component.html")
+})
+export class Export3DMapButtonComponent {
+	private fileName: string
+	private files: FileState[]
+	exporter = new STLExporter()
```
Comment (ce-bo) [inline visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts:21]: Can this property be private?

File: visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	constructor(@Inject(Store) store: Store) {
+		store.select(accumulatedDataSelector).subscribe(accumulatedData => {
+			this.fileName = accumulatedData.unifiedFileMeta?.fileName
+		})
+		store.select(filesSelector).subscribe(files => {
+			this.files = files
+		})
+	}
+
+	downloadStlFile() {
```
Comment (ce-bo) [inline visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts:32]: I would add some tests for this logic. Some of my ideas: Propably for edge cases like Empty ThreeMesh, One Selected File, Multiple Selected Files, Delta File. And I would test that the right stl file will be downloaded if the mode changes or another map was uploaded.

**Review by ce-bo — COMMENTED** [review]
The code is LGTM! :-) 
Please add some tests and adjust the changelog. Thanks!

File: visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts (hunk 3 above)
Comment (IhsenBouallegue) [reply visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts:32]: I would love some help with that. I tried but didn't succeed, as I was constantly stuck.

File: visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.scss [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+	.button-icon {
+		font-size: 16px;
+		line-height: 1.3;
+	}
+	span {
+		padding: 0;
+		margin: 0;
+	}
+	.mat-mini-fab .mat-button-wrapper {
+		height: 25px;
```
Comment (shaman-apprentice) [inline visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.scss:22]: width and height are redundant with above `.mat-mini-fab` and can be removed here, aren't they?

File: visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	selector: "cc-export-threed-map-button",
+	template: require("./export3DMapButton.component.html")
+})
+export class Export3DMapButtonComponent {
+	private fileName: string
+	private files: FileState[]
+	private exporter = new STLExporter()
+
+	constructor(@Inject(Store) store: Store) {
+		store.select(accumulatedDataSelector).subscribe(accumulatedData => {
```
Comment (shaman-apprentice) [inline visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts:24]: we must unsubscribe, if we subscribe manually

File: visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts (hunk 3 above)
Comment (shaman-apprentice) [reply visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts:32]: The download doesn't depent on the CodeCharta state, does it? As far as I understand it, it takes the Three.js model and generates a binary for the 3d printer out of it. Therefore we don't need to test different CodeCharta states like multiple or delta mode...

I would write a test that the FileDonloader gets called. The FileDownlaoder then could be tested through pure unit tests, I hope.

File: visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.scss (hunk 4 above)
Comment (IhsenBouallegue) [reply visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.scss:22]: Resolved in the next push

File: visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts (hunk 3 above)
Comment (ce-bo) [reply visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts:34]: Ah, I see. You are right!
@shaman-apprentice Does unifiedFileMeta mean that it is a combination of all selected files?

Comment (shaman-apprentice) [reply visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts:34]: yes. But I am not to sure about the naming and calculation - it is extracted by one of the services into a selector, as it is purely dependend from store state and can then easily be used by Angular components (and has better separations of concers, as the services do too much anyway)

Comment (shaman-apprentice) [reply visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts:34]: We added a test which verifies the template and controller binding. 

The download logic is still untested. It uses mostly external libs which are awkward to test with current design. We discussed a better approach, which is possible after Angular migration. Current implemented approach is still better than our "normal" Controller tests. According to the boy scouts rule I am OKish with it for now  (although the rule only kind of applies, as it is new code).

**Review by shaman-apprentice — APPROVED** [review]
Looks OK to me

File: visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts (hunk 3 above)
Comment (ce-bo) [reply visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts:34]: Personally, I would try to spy on the STLExporter:parse() and the FileDownloader:downloadData() function. Then, I would just test that the file name is set correctly.

Comment (ce-bo) [reply visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts:34]: I tested locally to download the STL file for 5 selected maps in multiple mode. The resulting file name is very long: `file_aggregation_of_sample1_and_sample2_and_codecharta_2021-12-16_14-52_and_junit4.source.modified_and_codecharta_edges_and_codecharta_analysis_2021-12-18_12-45.stl`. 

@BridgeAR Do you have any requirement for that?

**Review by ce-bo — APPROVED** [review]
LGTM! Thanks for completing the Tests. Would you please Update the Name of the new Test? Something like „should Download stl file Witz proper Name“ Would be More appropriate.
